packages/rest/src/responses.ts

```diff
@@ -0,0 +1,294 @@
+  responseModel: Type
+) {
+  // Get explicity defined status codes
+  const statusCodes: Array<string | typeof NoStatusCode> = getResponseStatusCodes(
+    program,
+    responseModel
+  );
+
+  // Get explicitly defined content types
+  const contentTypes = getResponseContentTypes(program, responseModel);
+
+  // Get response headers
+  const headers = getResponseHeaders(program, responseModel);
+
+  // Get explicitly defined body
+  let bodyModel = getResponseBody(program, responseModel);
+  // If there is no explicit body, it should be conjured from the return type
+  // if it is a primitive type or it contains more than just response metadata
+  if (!bodyModel) {
+    if (responseModel.kind === "Model") {
+      if (isIntrinsic(program, responseModel)) {
+        bodyModel = responseModel;
+      } else {
+        const isResponseMetadata = (p: ModelTypeProperty) =>
+          isHeader(program, p) || isStatusCode(program, p);
+        const allProperties = (p: ModelType): ModelTypeProperty[] => {
+          return [...p.properties.values(), ...(p.baseModel ? allProperties(p.baseModel) : [])];
+        };
+        if (allProperties(responseModel).some((p) => !isResponseMetadata(p))) {
+          bodyModel = responseModel;
+        }
+      }
+    } else {
+      // body is array or possibly something else
+      bodyModel = responseModel;
+    }
+  }
+
+  // If there is no explicit status code, check if it should be 204
+  if (statusCodes.length === 0) {
+    if (bodyModel === undefined || isVoidType(bodyModel)) {
+      statusCodes.push("204");
+    } else {
+      statusCodes.push(NoStatusCode);
+    }
+  }
+
+  // If there is a body but no explicit content types, use application/json
+  if (bodyModel && !isVoidType(bodyModel) && contentTypes.length === 0) {
+    contentTypes.push("application/json");
+  }
+
+  // Assertion: bodyModel <=> contentTypes.length > 0
+
+  // Put them into currentEndpoint.responses
+
+  for (const statusCode of statusCodes) {
+    // the first model for this statusCode/content type pair carries the
+    // description for the endpoint. This could probably be improved.
+    const response: HttpOperationResponse = responses[statusCode] ?? {
+      statusCode: statusCode === NoStatusCode ? undefined : statusCode,
+      type: responseModel,
+      description: getResponseDescription(program, responseModel, statusCode),
+    };
+
+    // check for duplicates
+    if (response.body) {
+      for (const contentType of contentTypes) {
+        if (response.body[contentType]) {
+          reportDiagnostic(program, {
+            code: "duplicate-response",
+            format: { statusCode: statusCode.toString(), contentType },
+            target: responseModel,
+          });
+        }
+      }
+    }
+
+    if (Object.keys(headers).length > 0) {
+      response.headers ??= {};
+
+      // OpenAPI can't represent different headers per content type.
```

Comment:
This seems like OA-specific logic because OA chooses not to allow the user to specify different headers for different content types. In general I'm concerned about conflating general cadl reflection logic with OA-specific logic. Is that a concern you share?

One way to think about this problem is, cadl-rest lets you reflect the cadl declarations into useful records, but in an explicitly non-lossy way. The headers violate the lossyness principle, and I'm not sure the return values are generally useful but rather geared toward slamming into openapi output.

Reply:
yes didn't think about that one, I do agree that the http library should be completely unaware of openapi and not depend on any of its concept, actually having issue with the default status code with this change. Fails a test where there is 2 response without statuscode but one has `@defaultResponse`